### DISCUSSION TITLE
Week 3 Code Review

### DISCUSSION THREAD

**Discussion by GalexY727** [Patribots4738/Crescendo2024#59] [Show and tell]
This has been a great week. 162 contributions on a day without a meeting is better than any FRC team I know, (even [Mechanical Advantage](https://thebluealliance.com/team/6328/2023)), and I am very proud of all the things that are being accomplished. This thread will be the start of a weekly trend I intend to do to help with documenting our code base as it goes through all 7 stages of ~~grief~~ issues.
This isn't a me-only thing. When a big change is made, please add it to this thread. Think of this discussion as a "Change log" or "Release notes" that will make our discussion at the end of the week easy to go through. I'll try to start, but if I missed something please add it!
## The big discussion: #22 
At the start of the week, the overall structure of the code was "decided" and mapped out for the first time, as seen in #22's drawing board. This set the stage for the rest of the week, as a _bunch_ of subsystems were made (cough cough) #24 #30 #29 #26 #31, and just about all of them are almost done or in review. They have been merged into #27's branch, with the exception of #30. I'm not sure why this is, but the TriggerWheel still has its own branch. Soon™️ it will be merged into #27, surely. After #27 is all done and dusted, to main it goes! Please remember that **no commits should ever be done directly to main**, as that opens the possibility of breaking our commit history which makes changes harder to revert.
### PieceControl Command: (#27)
Lots of talking was done as to how this file should be structured, but I would like to highlight #55 especially. This ~~led~~ (will lead?, is leading?) to the removal of #28, as it can just be integrated into #26. Clean code, happy aleg.
### Limelight #41 #39 #36 #53 #38  (why are there SO MANY)
Limelight was added, removed, added, removed, added, removed, added, removed, and added again. It should be *done* flip-flopping, and it will live on main, and if there are issues a PR will be made to fix it. 

That's all I have time for, please add more if you deem necesary.

Issues mentioned:
- [ ] #22 
- [ ] #24 
- [x] #30 
- [x] #29 
- [x] #26 
- [x] #31 
- [ ] #27 
- [x] #28 
- [ ] #55 
- [x] #41 
- [x] #39 
- [x] #36 
- [x] #53 
- [x] #38

**Comment by jacueblol** [comment]
This was very well written !! Great few weeks everyone! I'm very happy with how well everything has gone thus far. 
I'd like to finish up reviewing everything that was missed... (hopefully)

## PieceControl now uses params
Not only was #27 created and worked on, but was also modified to use parameters for the subsystems (#48) Because of this issue, The Discussion™️  was also made and worked on.

## Soft Inversion created and In Progress
#56 and #62  solve the soft inversion neo issue for a better way to invert the neo.

## Discord Bot
The discord bot was also created to better understand what was happening on GitHub. #18 

## Resource Link was created
#37 

## Pigeon now works!!
After #15 #14 we discovered the problem behind the pigeon, which was great!

## Limelight
Limelight is now working for tags and for vision #32 #7 #16 Just a few more final tests

# What's next?
Well, first of all, let me say again how proud of everyone I am. This was/is a huge project that flew by a lot faster than I expected. You all did a great job, and I'm proud of how far you have come, even from knowing little Java. 

We are still working on merging and resolving the issues for the final piece control component. Once that is complete, everything will be closed. (here is a list of issues for the subsystems that you all have worked on)
- [x] #25 
- [x] #29  
- [x] #31 
- [x] #26 
- [ ] #24 
- [x] #30 
- [x] #28 
- [x] #33 
- [ ] #27 
After all of that is complete, hopefully, we will be able to test on an actual bot and test some autos as well.
Please make sure to make some autos if you are feeling very robot. #2 

Tina is also working on a branch right now for #13 #9. Which will be tested and finished later.

If you want to pick up or suggest an issue or feature yourself, don't be scared to do so in the issues tab. Again I'm all very proud of where you all are at right now, I'm so happy with this week. Again always keep learning and happy coding! 

-Jacob

**Comment by GalexY727** [comment]
- [x] #65 
- [x] #68 
Important to mention ^
